internal/jobmanager/jobmanager.go

```diff
@@ -0,0 +1,497 @@
+// Provide an input key generation function to synchronize at the input item level.
+type Job[InputType any, OutputType any] struct {
+	// The user level options for a Job.
+	*JobOptions
+
+	// The number of workers in the parallel pool.
+	PoolSize int
+
+	// An optional key to lock on.
+	LockKey string
+
+	// A sorted list of items that need work.
+	WorkItems []InputType
+
+	// An optional function to retrieve existing records that should not be processed.
+	// Returns a list of processed records, remaining items, and an error.
+	RetrieveFunc func([]InputType) ([]OutputType, []InputType, error)
+
+	// An optional function to retrieve a record based on a single work item.
+	SingleRetreiveFunc func(InputType) (OutputType, bool, error)
+
+	// An optional function to store the result.
+	StoreFunc func([]OutputType) error
+
+	// An optional function to remove existing records from storage.
+	RemoveStorageFunc func([]InputType) error
+
+	// A function to transform work items into results.
+	// Returns a result, the time of computation, and an error.
+	WorkFunc func(InputType) (OutputType, error)
+
+	// An optional function to get the locking key for an individual work item.
+	WorkItemKeyFunc func(InputType) string
+
+	// An optional function to store stats on how long the job took.
+	StatFunc func(int64)
+}
+
+// JobOutput is the result of a call to Job.Run().
+// It contains a system error, a map of work errors, the result items,
+// the remaining items, the total run time, and a channel that signals execution is complete.
+// If the context is cacelled while running a job,
+// the returned JobOutput will be empty.
+type JobOutput[InputType any, OutputType any] struct {
+	// An error for Job.Run().
+	Error error
+
+	// A map of errors encountered while running the job.
+	// The key of the error will be the index of the item
+	// that failed in RemainingItems.
+	WorkErrors map[int]error
+
+	// The list of results.
+	ResultItems []OutputType
+
+	// The list of tems that still need work.
+	RemainingItems []InputType
+
+	// The total computation time.
+	RunTime int64
+
+	// Signals the job is complete.
+	Done <-chan any
+
+	// A UUID to synchronize operations.
+	JobID string
+}
+
+func (this *Job[InputType, OutputType]) Validate() error {
+	if this == nil {
+		return fmt.Errorf("Job is nil.")
+	}
+
+	if this.WorkFunc == nil {
+		return fmt.Errorf("Job cannot have a nil work function.")
+	}
+
+	if this.PoolSize <= 0 {
+		return fmt.Errorf("Pool size must be positive, got %d.", this.PoolSize)
+	}
+
+	return this.JobOptions.Validate()
+}
+
+func (this *JobOptions) Validate() error {
+	if this == nil {
+		return fmt.Errorf("Job options are nil.")
+	}
+
+	if this.Context == nil {
+		this.Context = context.Background()
+	}
+
+	return nil
+}
+
+func (this JobOutput[InputType, OutputType]) GetError() error {
+	lockmanager.ReadLock(this.JobID)
+	defer lockmanager.ReadUnlock(this.JobID)
```

Comment:
Since you have `ReturnCopy` (which may be renamed),
I don't think you need any of this locking behavior.

Let me know if I am missing something, but either:
 - `ReturnCopy` is true and the caller has their own copy they can use.
 - `ReturnCopy` is false, and the caller should wait on the done chan.